Fix API Installer to generate unique Operation IDs

mbohlool opened the pull request:
OperationIDs generated by api_installer.go may not be unique and that causes long auto-generated function names in swagger generated clients. This is a step toward making those IDs unique.

Reference Issue kubernetes/features#53 and #13414

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33337)

<!-- Reviewable:end -->

mbohlool committed:
Fix API Installer to generate unique Operation IDs

mbohlool committed:
Generated files

mbohlool commented:
Note: The PR labeled as "size/L" because of the generated code. It is not actually that big (first commit is the manual one).

lavalamp commented on api/swagger-spec/v1.json:
```diff
@@ -15402,7 +15402,7 @@
       "method": "GET",
       "summary": "connect GET requests to proxy of Service",
-      "nickname": "connectGetNamespacedServiceProxy",
+      "nickname": "connectGetNamespacedServiceProxyWithPath",
```
It's unfortunate that we have to split out every request to the proxy separately by verb. :/

mbohlool replied:
it is outside of the scope of this change, I will let it pass for this PR but will think about it.

lavalamp commented on api/swagger-spec/v1.json:
```diff
@@ -13193,7 +13193,7 @@
       "method": "GET",
       "summary": "watch individual changes to a list of Secret",
-      "nickname": "watchNamespacedSecretList",
+      "nickname": "watchNamespacedSecretListForAllNamespaces",
```
I'm not sure how critical the first "Namespaced" is. E.g., maybe (watchNamespacedSecretList, watchNamespacedSecretListForAllNamespaces) could instead be (watchSecretList, watchSecretListForAllNamespaces)?

(It's in there to distinguish between cluster and namespace scoped resources. I'm not sure if we have any with colliding names any more. @deads2k?)

lavalamp replied:
If it is important to keep the Namespaced at the beginning, maybe the second version could be watchAllNamespacesSecretList? That would keep the length down a little bit, although it's a bit less natural english.

mbohlool replied:
Removing first "Namespaced" make more sense to me. At some point I thought about it, but I am not sure why I didn't go with it.

lavalamp reviewed (commented):
The other thing to think about is if we should include group and version--that would future-proof these names a bit.

lavalamp commented:
I think my comments are optional :)

mbohlool commented:
About adding group and version, I think clientset would prevent that. We are generating one client per group/version. Do you suggest we have one client for all and add group/version to the name?

smarterclayton commented on pkg/apiserver/api_installer.go:
```diff
@@ -47,10 +47,11 @@ type APIInstaller struct {
+	Path          string               // The path of the action
+	Params        []*restful.Parameter // List of parameters associated with the action.
+	Namer         ScopeNamer
+	allNamespaces bool // true iff the action is namespaced but works on aggregate result for all namespaces
```
Should be public, like other fields on this type.

mbohlool replied:
Done.

smarterclayton commented:
One comment from me

lavalamp commented:
Ah, good point re: group/version. I forgot we were doing each client separately.

mbohlool commented:
I've addressed all comments. I am still thinking about group/version (for another PR). Please LGTM this one if it looks good.

For group/version, I was thinking about supporting a customization method in spec generation code and use it only for root swagger spec to add group/version to the name of operation. I will send another PR for that.

lavalamp commented:
LGTM. Customization for the universal doc sounds good, too.

mbohlool commented:
Only change was gofmt. applying LGTM again.
